### ISSUE TITLE
Use ProcessBuilder

### ISSUE THREAD

**Issue by danlooo** [Open-EO/openeo-julia-client#22]
The python client of openEO allows to use normal Python operators resulting in clean and concise code:

```python
bbox = {
        "west": 6.8371137, "south": 50.560007,
        "east": 6.8566699, "north": 50.5647147,
        "crs": "EPSG:4326"
    }
cube = connection.load_collection("TERRASCOPE_S2_TOC_V2",bands=['B04','B08'])\
    .filter_temporal("2017-10-10", "2017-10-30")\
    .filter_bbox(**bbox)

red = cube.band("B04")
nir = cube.band("B08")
ndvi = (nir - red) / (red + nir)
```

The Operators (+,-,/) are implemented in the class [`ProcessBuilder`](https://github.com/Open-EO/openeo-python-client/blob/554d573339e0a8ac8adbe55c6d617d4671e280cf/docs/api-processbuilder.rst#id7). 
This is based on the builder design pattern in OOP that is helpful in abstract type conversions.

Can we use this in the Julia client?

**Comment by danlooo** [comment]
OOP design patterns are obsolete in Julia and can't be directly used. However, multiple dispatch might be useful in processing:

1. Band arithmetic, e.g.  `cube1.red - cube1.nir`: Binary operators of types (band,band) or (band, number)
2. Reducer functions, e.g. `cube1.reduce_dimension(reducer =  x -> cos(x.mean()) + 1, dimension="t")`
3. Apply a function to each element of a data cube, e.g. `cube1.apply(x -> x /255)`

**Comment by danlooo** [comment]
Create and dispatch on a Node variable, analog to ProcessBuilder:

```julia
import Base: +, -, *, /, cos
struct Node
    op
    children::Vector
end
Node(x) = Node("create", [x]) # leaf

+(x::Node, y::Node) = Node("add", [x, y])
*(x::Node, y::Real) = Node("multiply", [x, Node(y)])
cos(x::Node) = Node("cos", [x])
g = (cos(Node(1)) + Node(3)) * 5
```

**Comment by meggart** [comment]
For functional processing patterns I think it would be nice to also rather overload Julias base functions as an alternative to the rather OOP-ish `cube1.reduce_dimension(...)`. E.g. we could define `Base.map(f, cube::Node)=cube.apply(f)` and provide similar overloads for `sum`, `mean`, `median` etc so that users could run code like

````julia
step2 = map(x -> x / 255, step1)
step3 = sum(step2, dims="t")
````

**Comment by danlooo** [comment]
Simple algebra is added in  [8d21293](https://github.com/Open-EO/openeo-julia-client/commit/8d21293c25a0bb13367f7355216e82889dcc015d). There was no need for promotion. Instead, the type DataCube was introduced with functions e.g. `+(cube::DataCube,number::Real)`. This type just contains the graph of openEO process calls that will create it.

**Comment by danlooo** [comment]
Closed, see also #23  and #24
